### DISCUSSION TITLE
Pyg explaination for edge classification model.

### DISCUSSION THREAD

**Discussion by arthurserres** [pyg-team/pytorch_geometric#8488] [Q&A] [answered]
Hello all, I recently discovered the Explain module of torch geometric. I want to use it to explain an edge classification model. Unfortunately, I have two main issues:

1 - Is this framework supported by explain, maybe can you tell me where I can find an example for an edge classification model ? 

2 - My graph is huge and I usually need to use loaders for training and evaluation (Because of memory limit). Some tests with explaination made my kernel crash, is there a way to use loaders, or is it a problem that you know how to adress ?

To be 100% clear, here is the project I try to make explainations about: [Multi-GNN](https://github.com/IBM/Multi-GNN)

Many thanks by advance.

Best,

Arthur

**Comment by xnuohz** [comment]
1. For GNN Explainer and Link Prediction, you may check this one: https://github.com/pyg-team/pytorch_geometric/blob/master/examples/explain/gnn_explainer_link_pred.py
2. For DataLoader, you can try here: https://pytorch-geometric.readthedocs.io/en/latest/modules/loader.html

**Comment by arthurserres** [comment]
1 - I was refering ot edge binary classification, not link prediction, any guess for that ?

2 - I was refering to loaders for the explain class, namely how to use loaders with explaination. My datas are too large to be charged in one time on the computer. Any idea about how to use loaders with explain. Or simply to manage memory issues with explain ?

Thanks by advance.

Arthur

**Reply by rusty1s** [reply, marked as answer]
1. Yes, both edge classification and link prediction is supported. In general, any task is supported as long as the output can be mapped to a given ground-truth.
2. This should work as well. You should be able to iterate over your graph and explain it in batches.

**Reply by arthurserres** [reply]
Thank you very much for the clarification, that is very nice of you. By any chance, do you have example of one of or both cases ?

**Reply by rusty1s** [reply]
This example here shows how to use `Explainer` in edge regression tasks: https://github.com/pyg-team/pytorch_geometric/blob/master/examples/explain/captum_explainer_hetero_link.py. We are currently missing an example within data loaders, but it really just comes down to calling `explainer` in each mini-batch iteration.

**Reply by arthurserres** [reply]
Thanks a lot for this answer. I will have a deeper look in it now. I will close the discussion and get back to you if needed !

Thanks a lot for your job :)

**Reply by arthurserres** [reply]
I looked again at this loader question and would like to check with you what should be done. What I understand from your answer is that we simply need to run something like 

```
for batch in tqdm.tqdm(loader):  
    with torch.no_grad():
                batch.to(device)
                explanation = explainer(
                batch.x,
                batch.edge_index,
                index=index,
                edge_attr=batch.edge_attr
                )
```

and then at the end of the process, explaination will contain the explaination for the full graph, is that correct ?

**Reply by rusty1s** [reply]
Partially. `explanation` will contain the explanation for the current mini-batch / or for the nodes in the mini-batch specified in `index`. As such, you would need to collect `explanation` per mini-batch inside a list.

**Reply by arthurserres** [reply]
I see, thanks for the answer! But for instance, does the last explaination take into account the full graph when performing explaination or only the graph included in the mini-batch ? Because in the latter case, the predictions that are used are not the predictions of the full model right ? (indeed let's say we want do predict the label of some edge, the prediction depends on the features but also on the graph so will change depending whether we consider the full graph or only the graph in the mini-batch) ?

**Reply by rusty1s** [reply]
The explanation will only include the input it is based on, which in this case, is the mini-batch/subgraph returned by the loader. Does this answer your question?

**Reply by arthurserres** [reply]
I think so yes, so there is no manner to produce an explaination using the full graph if the full graph is too large to be represented (in one time) in the memory ?

**Reply by rusty1s** [reply]
No, I wouldn't know how this could be possible.

**Reply by arthurserres** [reply]
Indeed it looks hard, thanks a lot for the answer :)
